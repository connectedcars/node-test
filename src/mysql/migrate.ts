import fs from 'fs'
import mysql from 'mysql'
import util from 'util'

import { MySQLClient } from './mysql-client'
import { dumpDatabase } from './mysqld-utils'

const readdir = util.promisify(fs.readdir)
const readFile = util.promisify(fs.readFile)
const mkdirAsync = util.promisify(fs.mkdir)
const existsAsync = util.promisify(fs.exists)

export interface MigrationRow {
  timestamp: string
  name: string
}

export interface Migration {
  path: string
  timestamp: string
  name: string
  sql: string
}

export interface MigrationResult {
  [schema: string]: SchemaMigrationResult
}

export interface SchemaMigrationResult {
  [database: string]: Migration[]
}

export interface MigrateOptions {
  mysqlClient: MySQLClient
  migrationsPaths?: string[]
  cachePaths?: string[]
  ignoreCache?: boolean
}

export class Migrate {
  private mysqlClient: MySQLClient
  private migrationsDir = ''
  private cachePaths: string[] = []
  private initPromise: Promise<void>
  private schemaFolders!: string[]
  private databaseMap!: { [key: string]: string[] }
  private basePool!: mysql.Pool
  private ignoreCache: boolean
  private timings: string[] = []
  private migrationsPaths: string[]

  public constructor(options: MigrateOptions) {
    this.mysqlClient = options.mysqlClient
    this.migrationsPaths = options.migrationsPaths || ['./migrations']
    this.cachePaths = options.cachePaths || ['./cache']
    this.ignoreCache = options.ignoreCache || false
    this.initPromise = this.init()
    this.initPromise.catch(() => {
      // Ignore so we don't et unhandled promise rejection
    })
  }

  public async init(): Promise<void> {
    // Find databases.json
    for (const migrationPath of this.migrationsPaths) {
      if (await existsAsync(`${migrationPath}/databases.json`)) {
        this.migrationsDir = migrationPath
        break
      }
    }
    if (!this.migrationsDir) {
      throw new Error(`databases.json could not be found in any of the migration paths`)
    }
    await this.saveTiming<void>('loadDatabaseJson', async () => {
      const [databasesJSON, schemaFolders] = await Promise.all([
        readFile(`${this.migrationsDir}/databases.json`, 'utf8'),
        readdir(this.migrationsDir)
      ])
      this.databaseMap = JSON.parse(databasesJSON)
      this.schemaFolders = schemaFolders.filter(s => s in this.databaseMap)
    })
    this.basePool = await this.mysqlClient.getConnectionPool('mysql')
  }

  public async getTimings(): Promise<string[]> {
    await this.initPromise // Make sure init has finished
    return this.timings
  }

  public async cleanup(): Promise<void> {
    await this.initPromise
    for (const schemaFolder of this.schemaFolders) {
      for (const database of this.databaseMap[schemaFolder]) {
        await this.mysqlClient.dropDatabase(database)
      }
    }
  }

  public async cacheSchemas(): Promise<void> {
    if (this.cachePaths.length === 0) {
      throw new Error('No cache paths defined')
    }
    await mkdirAsync(this.cachePaths[0], { recursive: true, mode: 0o777 })
    for (const schemaFolder of this.schemaFolders) {
      await dumpDatabase(
        this.mysqlClient.options.port || 0, // TODO: Make this a lot prettier
        this.databaseMap[schemaFolder],
        `${this.cachePaths}/${schemaFolder}.sql`
      )
    }
  }

  public async migrate(until?: string): Promise<MigrationResult> {
    await this.initPromise
    const result: MigrationResult = {}
    const promises: Array<Promise<SchemaMigrationResult>> = []
    const lockConnection = await this.mysqlClient.getConnection('mysql')
    try {
      await this.mysqlClient.takeLock(lockConnection, 'migrate')
      for (const schemaFolder of this.schemaFolders) {
        promises.push(this.migrateSchema(schemaFolder, until))
      }
      const migrations = await this.saveTiming('migrateAllSchema', Promise.all(promises))
      // Populate the MigrationResult structure
      for (const schemaFolder of this.schemaFolders) {
        result[schemaFolder] = migrations.shift() || {}
      }
      return result
    } finally {
      await this.mysqlClient.closeConnection(lockConnection)
    }
  }

  public async migrateSchema(schemaFolder: string, until?: string): Promise<SchemaMigrationResult> {
    // Find the first cache file that matches
    let selectCacheFile = ''
    for (const cachePath of this.cachePaths) {
      const possibleCacheFile = `${cachePath}/${schemaFolder}.sql`
      if (await existsAsync(possibleCacheFile)) {
        selectCacheFile = possibleCacheFile
        break
      }
    }

    // Find out if we should skip cache if all database already exists for this schemaFolder
    if (!this.ignoreCache && selectCacheFile) {
      const existingDatabases = await this.mysqlClient.queryArray<string>(
        this.basePool,
        `
          SELECT SCHEMA_NAME as \`name\`
          FROM information_schema.SCHEMATA;
        `
      )
      const skipCache = this.databaseMap[schemaFolder].some(d => existingDatabases.includes(d))
      if (!skipCache) {
        const cacheData = await readFile(selectCacheFile, 'utf8')
        try {
          await this.saveTiming(`applyCacheFile(${selectCacheFile})`, this.mysqlClient.query(this.basePool, cacheData))
        } catch (e) {
          throw new Error(`Failed applying cache file ${selectCacheFile}: ${e}`)
        }
      }
    }

    const result: SchemaMigrationResult = {}
    const migrations = await this.readMigrations(schemaFolder)
    const migrationPromises: Array<Promise<Migration[]>> = []
    for (const database of this.databaseMap[schemaFolder]) {
      migrationPromises.push(this.migrateDatabase(database, migrations, until))
    }
    const dbMigrationsResults = await this.saveTiming(
      `applyAllMigrations(${schemaFolder})`,
      Promise.allSettled(migrationPromises)
    )

    // Populate the SchemaMigrationResult structure
    for (const database of this.databaseMap[schemaFolder]) {
      const migrationResult = dbMigrationsResults.shift()
      if (migrationResult?.status === 'fulfilled') {
        result[database] = migrationResult.value
      } else if (migrationResult?.status === 'rejected') {
        throw migrationResult.reason
      }
    }
    return result
  }

  public async migrateDatabase(database: string, migrations: Migration[], until?: string): Promise<Migration[]> {
    // Create the database and migration if they do not exist
    await this.mysqlClient.query(
      this.basePool,
      `
        CREATE DATABASE IF NOT EXISTS \`${database}\` CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci;
        CREATE TABLE IF NOT EXISTS \`${database}\`.Migrations (
          id INT(11) UNSIGNED NOT NULL AUTO_INCREMENT,
          timestamp VARCHAR(255) NOT NULL,
          name VARCHAR(255) NOT NULL,
          createdAt TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
          PRIMARY KEY (id)
        );
      `
    )

    const pool = await this.mysqlClient.getConnectionPool(database)

    // Fetch applied migrations
    const appliedMigrations = await this.mysqlClient.query<MigrationRow>(
      pool,
      'SELECT `timestamp`, `name` FROM `Migrations`'
    )

    const result: Migration[] = []
    for (const migration of migrations) {
      // Skip if the full migration has been applied
      if (appliedMigrations.find(v => v.timestamp === migration.timestamp && v.name === migration.name)) {
        continue
      }

      // https://www.mysqltutorial.org/mysql-comment/
      // Remove comments and split migrations into different statements except if they have LOCK or TRANSACTION in them
      const migrationStatements = migration.sql.match(/LOCK TABLES|START TRANSACTION/i)
        ? [migration.sql]
        : migration.sql
            .replace(/--\n|--\s[^\n]*|#[^\n]*|\/\*[^!].*?\*\//gs, '')
            .split(';')
            .filter(s => !s.match(/^\s*$/s))

      const connection = await this.mysqlClient.getConnectionFromPool(pool)
      try {
        for (const [i, migrationStatement] of migrationStatements.entries()) {
          const migrationStatementName = migrationStatements.length > 1 ? `${migration.name}:${i}` : migration.name
          if (appliedMigrations.find(m => m.timestamp === migration.timestamp && m.name === migrationStatementName)) {
            continue
          }
          await this.mysqlClient.query(connection, migrationStatement)
          await this.mysqlClient.query(connection, 'INSERT INTO `Migrations` (`timestamp`, `name`) VALUES (?, ?)', [
            migration.timestamp,
            migrationStatementName
          ])
          result.push(migration)

          if (until && until <= migration.timestamp) {
            break
          }
        }
      } catch (e) {
        throw new Error(`Failed applying migration "${migration.path}: ${e}"`)
      } finally {
        pool.releaseConnection(connection)
      }
    }

    return result
  }

  public async readMigrations(schemaPath: string): Promise<Migration[]> {
    const migrationsFiles = (await readdir(`${this.migrationsDir}/${schemaPath}`)).sort()
    const promises: Promise<Migration>[] = []
    for (const migrationFile of migrationsFiles) {
      promises.push(this.readMigration(`${schemaPath}/${migrationFile}`))
    }
    return Promise.all(promises)
  }

  public async readMigration(migrationFile: string): Promise<Migration> {
    const match = migrationFile.match(/\/([^_]+)_([^.]+\.sql)$/)
    if (!match) {
      throw new Error(`Migration file does not follow format: ${migrationFile}`)
    }
    const contents = await readFile(`${this.migrationsDir}/${migrationFile}`, 'utf8')
    const sql = contents.split(/\nEXIT/im).shift()
    if (!sql) {
      throw new Error(`Empty migration`)
    }
    return {
      path: migrationFile,
      timestamp: match[1],
      name: match[2],
      sql
    }
  }

  private async saveTiming<T>(name: string, wrap: (() => Promise<T>) | Promise<T>): Promise<T> {
    const start = process.hrtime()
    const result = typeof wrap === 'function' ? await wrap() : await wrap
    const diff = process.hrtime(start)
    this.timings.push(`${name} ${diff[0]}s ${diff[1] / 1000000}ms`)
    return result
  }
}
