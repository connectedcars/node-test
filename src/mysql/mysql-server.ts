import crypto from 'crypto'
import fs from 'fs'
import path from 'path'
import util from 'util'

const writeFileAsync = util.promisify(fs.writeFile)
const mkdirAsync = util.promisify(fs.mkdir)
const existsAsync = util.promisify(fs.exists)
const chmodAsync = util.promisify(fs.chmod)
const unlinkAsync = util.promisify(fs.unlink)

import { findFreePort } from '../net'
import {
  createTempDirectory,
  isDockerOverlay2,
  isPidRunning,
  readPidFile,
  stopPid,
  touchFiles,
  writePidFile
} from '../unix'
import {
  createMySQLDataCache,
  extractMySQLDataCache,
  generateMySQLServerConfig,
  getMySQLServerVersionString,
  initializeMySQLData,
  MySQLServerConfig,
  readPortFile,
  startMySQLd
} from './mysqld-utils'

export interface MySQLServerOptions {
  mysqlBaseDir?: string
  listenPort?: number
  mysqldPath?: string
  cachePath?: string
  myCnf?: MySQLServerConfig
  ignoreCache?: boolean
  ignoreCustomCache?: boolean
}

type InitStatus = 'unknown' | 'started' | 'initialized' | 'resumed'

const formatHrDiff = (what: string, diff: [number, number]): string => `${what} ${diff[0]}s ${diff[1] / 1000000}ms`

export class MySQLServer {
  private initStatus: InitStatus = 'unknown'
  private listenPort!: number
  private mysqlBaseDir!: string
  private timings: string[] = []

  private mysqldPath: string
  private myCnfCustom: MySQLServerConfig
  private mysqldPid!: number
  private initPromise: Promise<void>
  private options: MySQLServerOptions
  private cachePath: string
  private cleanInitializeDataTarGz!: string
  private customInitializeDataTarGz!: string
  private ignoreCache: boolean
  private ignoreCustomCache: boolean

  public constructor(options: MySQLServerOptions = {}) {
    this.options = options
    this.mysqldPath = options.mysqldPath || 'mysqld'
    this.cachePath = options.cachePath || './cache'
    this.myCnfCustom = options.myCnf || {}
    this.ignoreCache = options.ignoreCache || false
    this.ignoreCustomCache = (options.ignoreCache && options.ignoreCustomCache) || false
    this.initPromise = this.init()
    this.initPromise.catch(() => {
      // Do nothing as we will throw at later calls
    })
  }

  public async getTimings(): Promise<string[]> {
    await this.initPromise // Make sure init has finished
    return this.timings
  }

  public async getInitStatus(): Promise<InitStatus> {
    await this.initPromise // Make sure init has finished
    return this.initStatus
  }

  public async kill(sigKillTimeout = 5000): Promise<void> {
    await this.initPromise // Make sure init has finished
    await stopPid(this.mysqldPid, sigKillTimeout)
  }

  public async getListenPort(): Promise<number> {
    await this.initPromise // Make sure init has finished
    return this.listenPort
  }

  public async getConnectionUrl(database: string): Promise<string> {
    await this.initPromise
    return `mysql://root:@127.0.0.1:${this.listenPort}/${database}?charset=utf8mb4&multipleStatements=true`
  }

  public async getMysqlBaseDir(): Promise<string> {
    await this.initPromise // Make sure init has finished
    return this.mysqlBaseDir
  }

  public async saveAsCustomInitState(): Promise<void> {
    await this.initPromise // Make sure init has finished
    await createMySQLDataCache(this.mysqlBaseDir, this.customInitializeDataTarGz)
  }

  public async waitForStarted(): Promise<void> {
    await this.initPromise
  }

  private async init(): Promise<void> {
    // Generate unique cache key based on mysql version
    const mysqlVersion = await getMySQLServerVersionString(this.mysqldPath)
    const hash = crypto.createHash('sha1')
    const cacheCheckSum = hash.update(mysqlVersion).digest('hex') // TODO: Also add calculation for myCnf
    this.cleanInitializeDataTarGz = path.resolve(
      path.join(this.cachePath, `clean-initialization-data-${cacheCheckSum}.tar.gz`)
    )
    this.customInitializeDataTarGz = path.resolve(
      path.join(this.cachePath, `custom-initialization-data-${cacheCheckSum}.tar.gz`)
    )

    // Resume from or create new mysqlBaseDir
    let startingPidFile = ''
    if (this.options.mysqlBaseDir) {
      // TODO: Drop mysqlBaseDir if the version of configuration has changed.
      this.mysqlBaseDir = path.resolve(this.options.mysqlBaseDir)
      // Do simple locking to avoid race condition on startup from existing folder
      startingPidFile = path.join(this.mysqlBaseDir, 'starting.pid')
      await writePidFile(startingPidFile, 100)

      // Check if the mysqld is already running from this folder
      const mysqldPidFile = path.join(this.mysqlBaseDir, 'mysqld.pid')
      const pid = await readPidFile(mysqldPidFile)
      if (pid && (await isPidRunning(pid))) {
        const listenPort = await readPortFile(path.join(this.mysqlBaseDir, 'mysqld.port'))
        if (listenPort) {
          this.mysqldPid = pid
          this.listenPort = listenPort
          this.initStatus = 'resumed'
          return
        }
        // Kill the pid if we did not read a listen port
        await stopPid(pid, 3000)
      }
    } else {
      this.mysqlBaseDir = await createTempDirectory()
    }

    // Initialize mysql data
    let initialized = false
    if (!fs.existsSync(`${this.mysqlBaseDir}/data`)) {
      const myCnf: MySQLServerConfig = {
        mysqld: {},
        'mysqld-8.0': {}
      }
      if (process.getuid() === 0) {
        // Drop privileges if running as root
        myCnf.mysqld.user = 'mysql'
      }
      if (process.platform === 'darwin') {
        // Work around issue with mysql 5.7 running out of FD on macOSX: https://bugs.mysql.com/bug.php?id=79125
        // eslint-disable-next-line @typescript-eslint/camelcase
        myCnf.mysqld.table_open_cache = '250'
        // eslint-disable-next-line @typescript-eslint/camelcase
        myCnf.mysqld.open_files_limit = '800'
        // eslint-disable-next-line @typescript-eslint/camelcase
        myCnf.mysqld.max_connections = '500'

        // Set limits higher for 8.x
        // eslint-disable-next-line @typescript-eslint/camelcase
        myCnf['mysqld-8.0'].max_connections = '2000'
        // eslint-disable-next-line @typescript-eslint/camelcase
        myCnf['mysqld-8.0'].open_files_limit = '5000'
        // eslint-disable-next-line @typescript-eslint/camelcase
        myCnf['mysqld-8.0'].table_open_cache = '2000'
      }

      // Create base dir
      await mkdirAsync(path.join(this.mysqlBaseDir), { recursive: true, mode: 0o777 })
      await chmodAsync(this.mysqlBaseDir, '777')
      const config = generateMySQLServerConfig(this.mysqlBaseDir, { ...myCnf, ...this.myCnfCustom })
      await writeFileAsync(`${path.join(this.mysqlBaseDir, 'my.cnf')}`, config)

      // Make sure /files exists as it is used for LOAD
      await mkdirAsync(path.join(this.mysqlBaseDir, '/files'), { recursive: true, mode: 0o777 })
      await chmodAsync(path.join(this.mysqlBaseDir, '/files'), '777')

      // initialize mysql data folder
      const initializeTime = process.hrtime()
      if (!this.ignoreCustomCache && (await existsAsync(this.customInitializeDataTarGz))) {
        await extractMySQLDataCache(this.mysqlBaseDir, this.customInitializeDataTarGz)
      } else if (!this.ignoreCache && (await existsAsync(this.cleanInitializeDataTarGz))) {
        await extractMySQLDataCache(this.mysqlBaseDir, this.cleanInitializeDataTarGz)
      } else {
        await initializeMySQLData(this.mysqldPath, this.mysqlBaseDir)
        await mkdirAsync(this.cachePath, { recursive: true, mode: 0o777 })
        await createMySQLDataCache(this.mysqlBaseDir, this.cleanInitializeDataTarGz)
      }
      this.timings.push(formatHrDiff('initializeMySQLData', process.hrtime(initializeTime)))
      initialized = true
    } else if (await isDockerOverlay2()) {
      // Working around issue with docker overlay2
      await touchFiles(this.mysqlBaseDir as string)
    }

    // Find free port and start mysqld
    this.listenPort = this.options.listenPort ? this.options.listenPort : await findFreePort()
    const startTime = process.hrtime()
    this.mysqldPid = await startMySQLd(this.mysqldPath, this.mysqlBaseDir, [`--port=${this.listenPort}`])
    this.timings.push(formatHrDiff('startMySQLd', process.hrtime(startTime)))
    await writeFileAsync(`${path.join(this.mysqlBaseDir, '/mysqld.port')}`, this.listenPort)
    this.initStatus = initialized ? 'initialized' : 'started'
    if (startingPidFile) {
      await unlinkAsync(startingPidFile).catch(() => {
        /* Ignore */
      })
    }
  }
}
