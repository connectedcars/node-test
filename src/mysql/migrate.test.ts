import { Migrate } from './migrate'
import { MySQLClient } from './mysql-client'
import { MySQLServer } from './mysql-server'

async function time<T>(promise: Promise<T>): Promise<[T, number]> {
  const start = process.hrtime()
  const result = await promise
  const diff = process.hrtime(start)
  return [result, diff[0] * 1000000 + diff[1] / 1000]
}

describe('Migrate', () => {
  let mySqlClient: MySQLClient

  beforeAll(async () => {
    const mySqlServer = new MySQLServer({ mysqlBaseDir: 'mysql-context' })
    console.log(await mySqlServer.getTimings())
    mySqlClient = new MySQLClient({ port: await mySqlServer.getListenPort() })
  }, 30000)

  afterEach(async () => {
    await mySqlClient?.cleanup()
  })

  it('should migrate schema over two migration runs', async () => {
    const migrate = new Migrate({
      mysqlClient: mySqlClient,
      migrationsPaths: ['src/mysql/resources/migrations'],
      ignoreCache: true
    })
    await migrate.cleanup()
    const [migrationResultBefore, timingBefore] = await time(migrate.migrate('2020-04-02T165700'))
    console.log(timingBefore / 1000)
    const pool = await mySqlClient.getConnectionPool('my_test01')
    const columnsBefore = await mySqlClient.queryArray<string>(
      pool,
      `
        SELECT COLUMN_NAME AS 'column'
        FROM INFORMATION_SCHEMA.COLUMNS
        WHERE TABLE_SCHEMA ='my_test01' AND TABLE_NAME='VehicleInfo'
        ORDER BY 'column';
      `
    )
    expect(columnsBefore).toMatchSnapshot()
    expect(migrationResultBefore).toMatchSnapshot()

    const [migrationResultAfter, timingAfter] = await time(migrate.migrate('2020-04-02T165700'))
    console.log(timingAfter / 1000)
    const columnsAfter = await mySqlClient.queryArray<string>(
      pool,
      `
        SELECT COLUMN_NAME AS 'column'
        FROM INFORMATION_SCHEMA.COLUMNS
        WHERE TABLE_SCHEMA ='my_test01' AND TABLE_NAME='VehicleInfo'
        ORDER BY 'column';
      `
    )
    expect(columnsAfter).toMatchSnapshot()
    expect(migrationResultAfter).toMatchSnapshot()
  })

  it('should migrate schema creating cache and use this to restore state when migrating again', async () => {
    // Do initial migration and without using cache
    const initialMigrate = new Migrate({
      mysqlClient: mySqlClient,
      migrationsPaths: ['src/mysql/resources/migrations'],
      ignoreCache: true
    })
    await initialMigrate.cleanup()
    const migrationResultBefore = await initialMigrate.migrate()
    await initialMigrate.cacheSchemas()
    expect(migrationResultBefore).toMatchSnapshot()

    const cachedMigrate = new Migrate({
      mysqlClient: mySqlClient,
      migrationsPaths: ['src/mysql/resources/migrations']
    })

    await cachedMigrate.cleanup()
    const migrationResultAfter = await cachedMigrate.migrate()
    expect(migrationResultAfter).toMatchSnapshot()
  })

  it.skip('should migrate data repo to newest version', async () => {
    const migrate = new Migrate({ mysqlClient: mySqlClient, migrationsPaths: ['data/migrations'] })
    await migrate.cleanup()
    const [migrationResult, timingBefore] = await time(migrate.migrate())
    console.log(timingBefore / 1000)
  }, 60_000)
})
