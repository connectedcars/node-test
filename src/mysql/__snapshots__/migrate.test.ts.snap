// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migrate should migrate schema creating cache and use this to restore state when migrating again 1`] = `
Object {
  "test": Object {
    "my_test01": Array [
      Object {
        "name": "vehicle-info_add_table.sql",
        "path": "test/2020-04-02T165700_vehicle-info_add_table.sql",
        "sql": "CREATE TABLE \`VehicleInfo\` (
    \`id\` INT(11) UNSIGNED NOT NULL AUTO_INCREMENT,
    \`vin\` VARCHAR(17) NOT NULL,
    \`vendor\` VARCHAR(255) NOT NULL, -- VAG
    \`make\` VARCHAR(255) NOT NULL, -- Audi
    \`name\` VARCHAR(255) NOT NULL, -- Audi Q2 Sport
    \`year\` YEAR(4) NOT NULL, -- 2018

    \`createdAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    \`updatedAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,

    PRIMARY KEY (\`id\`),
    UNIQUE KEY \`vin\` (\`vin\`)
)
ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8mb4
COMMENT 'Basic vehicle information'
;
",
        "timestamp": "2020-04-02T165700",
      },
      Object {
        "name": "vehicle-info_allow_null.sql",
        "path": "test/2020-04-02T170100_vehicle-info_allow_null.sql",
        "sql": "ALTER TABLE \`VehicleInfo\` ADD \`model\` VARCHAR(255) NULL AFTER \`name\`;
",
        "timestamp": "2020-04-02T170100",
      },
    ],
    "my_test02": Array [
      Object {
        "name": "vehicle-info_add_table.sql",
        "path": "test/2020-04-02T165700_vehicle-info_add_table.sql",
        "sql": "CREATE TABLE \`VehicleInfo\` (
    \`id\` INT(11) UNSIGNED NOT NULL AUTO_INCREMENT,
    \`vin\` VARCHAR(17) NOT NULL,
    \`vendor\` VARCHAR(255) NOT NULL, -- VAG
    \`make\` VARCHAR(255) NOT NULL, -- Audi
    \`name\` VARCHAR(255) NOT NULL, -- Audi Q2 Sport
    \`year\` YEAR(4) NOT NULL, -- 2018

    \`createdAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    \`updatedAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,

    PRIMARY KEY (\`id\`),
    UNIQUE KEY \`vin\` (\`vin\`)
)
ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8mb4
COMMENT 'Basic vehicle information'
;
",
        "timestamp": "2020-04-02T165700",
      },
      Object {
        "name": "vehicle-info_allow_null.sql",
        "path": "test/2020-04-02T170100_vehicle-info_allow_null.sql",
        "sql": "ALTER TABLE \`VehicleInfo\` ADD \`model\` VARCHAR(255) NULL AFTER \`name\`;
",
        "timestamp": "2020-04-02T170100",
      },
    ],
    "my_test03": Array [
      Object {
        "name": "vehicle-info_add_table.sql",
        "path": "test/2020-04-02T165700_vehicle-info_add_table.sql",
        "sql": "CREATE TABLE \`VehicleInfo\` (
    \`id\` INT(11) UNSIGNED NOT NULL AUTO_INCREMENT,
    \`vin\` VARCHAR(17) NOT NULL,
    \`vendor\` VARCHAR(255) NOT NULL, -- VAG
    \`make\` VARCHAR(255) NOT NULL, -- Audi
    \`name\` VARCHAR(255) NOT NULL, -- Audi Q2 Sport
    \`year\` YEAR(4) NOT NULL, -- 2018

    \`createdAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    \`updatedAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,

    PRIMARY KEY (\`id\`),
    UNIQUE KEY \`vin\` (\`vin\`)
)
ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8mb4
COMMENT 'Basic vehicle information'
;
",
        "timestamp": "2020-04-02T165700",
      },
      Object {
        "name": "vehicle-info_allow_null.sql",
        "path": "test/2020-04-02T170100_vehicle-info_allow_null.sql",
        "sql": "ALTER TABLE \`VehicleInfo\` ADD \`model\` VARCHAR(255) NULL AFTER \`name\`;
",
        "timestamp": "2020-04-02T170100",
      },
    ],
  },
}
`;

exports[`Migrate should migrate schema creating cache and use this to restore state when migrating again 2`] = `
Object {
  "test": Object {
    "my_test01": Array [],
    "my_test02": Array [],
    "my_test03": Array [],
  },
}
`;

exports[`Migrate should migrate schema over two migration runs 1`] = `
Array [
  "createdAt",
  "id",
  "make",
  "name",
  "updatedAt",
  "vendor",
  "vin",
  "year",
]
`;

exports[`Migrate should migrate schema over two migration runs 2`] = `
Object {
  "test": Object {
    "my_test01": Array [
      Object {
        "name": "vehicle-info_add_table.sql",
        "path": "test/2020-04-02T165700_vehicle-info_add_table.sql",
        "sql": "CREATE TABLE \`VehicleInfo\` (
    \`id\` INT(11) UNSIGNED NOT NULL AUTO_INCREMENT,
    \`vin\` VARCHAR(17) NOT NULL,
    \`vendor\` VARCHAR(255) NOT NULL, -- VAG
    \`make\` VARCHAR(255) NOT NULL, -- Audi
    \`name\` VARCHAR(255) NOT NULL, -- Audi Q2 Sport
    \`year\` YEAR(4) NOT NULL, -- 2018

    \`createdAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    \`updatedAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,

    PRIMARY KEY (\`id\`),
    UNIQUE KEY \`vin\` (\`vin\`)
)
ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8mb4
COMMENT 'Basic vehicle information'
;
",
        "timestamp": "2020-04-02T165700",
      },
    ],
    "my_test02": Array [
      Object {
        "name": "vehicle-info_add_table.sql",
        "path": "test/2020-04-02T165700_vehicle-info_add_table.sql",
        "sql": "CREATE TABLE \`VehicleInfo\` (
    \`id\` INT(11) UNSIGNED NOT NULL AUTO_INCREMENT,
    \`vin\` VARCHAR(17) NOT NULL,
    \`vendor\` VARCHAR(255) NOT NULL, -- VAG
    \`make\` VARCHAR(255) NOT NULL, -- Audi
    \`name\` VARCHAR(255) NOT NULL, -- Audi Q2 Sport
    \`year\` YEAR(4) NOT NULL, -- 2018

    \`createdAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    \`updatedAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,

    PRIMARY KEY (\`id\`),
    UNIQUE KEY \`vin\` (\`vin\`)
)
ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8mb4
COMMENT 'Basic vehicle information'
;
",
        "timestamp": "2020-04-02T165700",
      },
    ],
    "my_test03": Array [
      Object {
        "name": "vehicle-info_add_table.sql",
        "path": "test/2020-04-02T165700_vehicle-info_add_table.sql",
        "sql": "CREATE TABLE \`VehicleInfo\` (
    \`id\` INT(11) UNSIGNED NOT NULL AUTO_INCREMENT,
    \`vin\` VARCHAR(17) NOT NULL,
    \`vendor\` VARCHAR(255) NOT NULL, -- VAG
    \`make\` VARCHAR(255) NOT NULL, -- Audi
    \`name\` VARCHAR(255) NOT NULL, -- Audi Q2 Sport
    \`year\` YEAR(4) NOT NULL, -- 2018

    \`createdAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP,
    \`updatedAt\` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,

    PRIMARY KEY (\`id\`),
    UNIQUE KEY \`vin\` (\`vin\`)
)
ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8mb4
COMMENT 'Basic vehicle information'
;
",
        "timestamp": "2020-04-02T165700",
      },
    ],
  },
}
`;

exports[`Migrate should migrate schema over two migration runs 3`] = `
Array [
  "createdAt",
  "id",
  "make",
  "model",
  "name",
  "updatedAt",
  "vendor",
  "vin",
  "year",
]
`;

exports[`Migrate should migrate schema over two migration runs 4`] = `
Object {
  "test": Object {
    "my_test01": Array [
      Object {
        "name": "vehicle-info_allow_null.sql",
        "path": "test/2020-04-02T170100_vehicle-info_allow_null.sql",
        "sql": "ALTER TABLE \`VehicleInfo\` ADD \`model\` VARCHAR(255) NULL AFTER \`name\`;
",
        "timestamp": "2020-04-02T170100",
      },
    ],
    "my_test02": Array [
      Object {
        "name": "vehicle-info_allow_null.sql",
        "path": "test/2020-04-02T170100_vehicle-info_allow_null.sql",
        "sql": "ALTER TABLE \`VehicleInfo\` ADD \`model\` VARCHAR(255) NULL AFTER \`name\`;
",
        "timestamp": "2020-04-02T170100",
      },
    ],
    "my_test03": Array [
      Object {
        "name": "vehicle-info_allow_null.sql",
        "path": "test/2020-04-02T170100_vehicle-info_allow_null.sql",
        "sql": "ALTER TABLE \`VehicleInfo\` ADD \`model\` VARCHAR(255) NULL AFTER \`name\`;
",
        "timestamp": "2020-04-02T170100",
      },
    ],
  },
}
`;
